src/box/txn_limbo.c

```diff
@@ -649,6 +662,63 @@ txn_limbo_read_demote(struct txn_limbo *limbo, int64_t lsn)
 	return txn_limbo_read_promote(limbo, REPLICA_ID_NIL, lsn);
 }
 
+/**
+ * Check that some synchronous transactions have gathered quorum and
+ * write a confirmation entry of the last confirmed transaction.
+ */
+static void
+txn_limbo_confirm(struct txn_limbo *limbo)
+{
+	if (limbo->is_in_rollback)
+		return;
+	struct txn_limbo_entry *e;
+	if (!limbo->confirmed_entry_is_valid) {
+		/* Try to find the next valid confirmed_entry value. */
+		e = rlist_next_entry(limbo->confirmed_entry, in_queue);
```

Comment:
I really like this idea, but there are some complications: `entry_to_confirm` does not always exist, unlike the previous one before it. And that's the only reason I did it this way.
Let's imagine a situation where there is the following limbo queue:
entry 1, entry2, ..., entry n - **all synchronous and all confirmed, there is nothing after entry n**.
Where will `entry_to_confirm` point? Okay, let it point to NULL. I can patch `txn_limbo_append`:
```diff
txn_limbo_append(struct txn_limbo *limbo, uint32_t id, struct txn *txn)
{
	//...
	struct txn_limbo_entry *e = region_alloc_object(&txn->region,
							typeof(*e), &size);
+	if(limbo->entry_to_confirm == NULL && txn_has_flag(txn, TXN_WAIT_ACK))
+		limbo->entry_to_confirm = e;
}
```
Hmm, maybe this can be done. Always, when I thought about this before, it seemed to me that we could come up with a use case where we would have to constantly run from the beginning along a large queue. But it seems not.


I'll try to rewrite it in this form and see if any pitfalls appear. It seems that even the rollback doesn’t really break anything.

Reply:
It seems to have worked. I wanted to ensure that if we had to iterate in search of entry_to_confirm, so that we would not have to go through the traversed entries a second time (so that each entry was guaranteed to be visited no more than once). And in this solution this is also achieved.

Reply:
Thanks for the thorough investigation!